src/methods.ts

```diff
@@ -226,6 +227,98 @@ export class Field extends AbstractMethodField {
   }
 }
 
+const opcodeSize: number[] = function() {
+  const table:number[] = [];
+  const layoutType = enums.OpcodeLayoutType;
+
+  table[layoutType.OPCODE_ONLY] = 1;
+  table[layoutType.CONSTANT_POOL_UINT8] = 2;
+  table[layoutType.CONSTANT_POOL] = 3;
+  table[layoutType.CONSTANT_POOL_AND_UINT8_VALUE] = 4;
+  table[layoutType.UINT8_VALUE] = 2;
+  table[layoutType.UINT8_AND_INT8_VALUE] = 3;
+  table[layoutType.INT8_VALUE] = 2;
+  table[layoutType.INT16_VALUE] = 3;
+  table[layoutType.INT32_VALUE] = 5;
+  table[layoutType.ARRAY_TYPE] = 2;
+  table[layoutType.WIDE] = 1;
+
+  return table;
+}();
+
+class TraceInfo {
+  pops: string[] = [];
+  pushes: string[] = [];
+  prefixEmit: string = "";
+  onErrorPushes: string[];
+
+  constructor(public pc: number, public jitInfo: JitInfo) {
+  }
+}
+
+class Trace {
+  private infos: TraceInfo[] = [];
+
+  constructor(public startPC: number, private code: Buffer, private method: Method) {
+  }
+
+  public addOp(pc: number, jitInfo: JitInfo) {
+    this.infos.push(new TraceInfo(pc, jitInfo));
+  }
+
+  public close(): Function {
+    if (this.infos.length > 1) {
+      const symbolicStack: string[] = [];
+      let symbolCount = 0;
+      let emitted = "";
+      for (let i = 0; i < this.infos.length; i++) {
+        const info = this.infos[i];
+        const jitInfo = info.jitInfo;
+
+        const pops = info.pops;
+        const normalizedPops = jitInfo.pops < 0 ? Math.min(-jitInfo.pops, symbolicStack.length) : jitInfo.pops;
+        for (let j = 0; j < normalizedPops; j++) {
+          if (symbolicStack.length > 0) {
+            pops.push(symbolicStack.pop());
+          } else {
+            const symbol = "s" + symbolCount++;
+            info.prefixEmit += `var ${symbol} = f.opStack.pop();`;
+            pops.push(symbol);
+          }
+        }
+
+        info.onErrorPushes = symbolicStack.slice();
+
+        const pushes = info.pushes;
+        for (let j = 0; j < jitInfo.pushes; j++) {
+          const symbol = "s" + symbolCount++;
+          symbolicStack.push(symbol);
+          pushes.push(symbol);
+        }
+
+      }
+
+      if (symbolicStack.length === 1) {
+        emitted += `f.opStack.push(${symbolicStack[0]});`;
+      } else if (symbolicStack.length > 1) {
+        emitted += `f.opStack.pushAll(${symbolicStack.join(',')});`;
+      }
+
+      for (let i = this.infos.length-1; i >= 0; i--) {
+        const info = this.infos[i];
+        const jitInfo = info.jitInfo;
+        emitted = info.prefixEmit + jitInfo.emit(info.pops, info.pushes, ""+i, emitted, this.code, info.pc, info.onErrorPushes, this.method);
+      }
+
+      // console.log(`Emitted trace of ${this.infos.length} ops: ` + emitted);
+      // f = frame, t = thread, u = util
+      return new Function("f", "t", "u", emitted);
```

Comment:
`new Function()` is nice because it does not have the same scoping issues that `eval` has, and is much more friendly for JavaScript JITs.

The problem with `new Function()` is that there's no way to add a line to the end of the body to force Chrome DevTools to surface the code as a particular document name, making it hard to debug. If you open Chrome DevTools with Doppio currently, you'll notice that classes and other eval'd code have a friendly name in the devtools.

I'm fine with keeping `Function`, but I wanted to note that in case you were wondering why I used `eval` elsewhere.